### PR TITLE
Keep skipped status

### PR DESCRIPTION
**Pull request by mgonnet**
## Problem description

If I skip a Scenario from a Before hook (like explained [here](https://github.com/cucumber/cucumber-js/blob/master/docs/support_files/hooks.md#skipping-in-a-before-hook)) but I also add an After hook, the execution of the After hook causes the status of the scenario to change from `SKIPPED` to `PASSED`

To see the issue on the [browser example](http://cucumber.github.io/cucumber-js/), paste the following code in the steps definitions
<details><summary>Step definitions</summary>
<p>

```javascript
// Cucumber and chai have been loaded in the browser
var setWorldConstructor = Cucumber.setWorldConstructor;
var Given = Cucumber.Given;
var When = Cucumber.When;
var Then = Cucumber.Then;
var Before = Cucumber.Before;
var After = Cucumber.After;
var expect = chai.expect;

///// World /////
//
// Call 'setWorldConstructor' with to your custom world (optional)
//

var CustomWorld = function() {
  this.variable = 0;
};

CustomWorld.prototype.setTo = function(number) {
  this.variable = parseInt(number);
};

CustomWorld.prototype.incrementBy = function(number) {
  this.variable += parseInt(number);
};

setWorldConstructor(CustomWorld);

///// Step definitions /////
//
// use 'Given', 'When' and 'Then' to declare step definitions
//

Before(function(){
    return 'skipped';
});

Given('a variable set to {int}', function(number) {
  this.setTo(number);
});

When('I increment the variable by {int}', function(number) {
  this.incrementBy(number);
});

Then('the variable should contain {int}', function(number) {
  expect(this.variable).to.eql(number)
});

After(function(){
    // some code
})
```
</p>
</details>

The output is the following:
image

But I would expect this:
image

## Motivation for the change:

As mentioned by @charlierudolph [here](https://github.com/cucumber/cucumber-js/issues/873#issuecomment-312775245) the After hooks must be executed even if the Scenario is skipped, but the Scenario should be marked as `SKIPPED` nevertheless in order to provide accurate metrics.

## The solution 
`shouldUpdateStatus` in `src/runtime/test_case_runner.js` now returns `false` if the current status of the test case is `SKIPPED`, with the only exception if the last step is `UNDEFINED`

In a more natural language: If there is at least one step SKIPPED, then the Scenario starts to skip the rest and is marked as SKIPPED, unless it has UNDEFINED steps

## Notes:
* Relates to #873 #912 
* Running `npm test` locally I get the following message:
<details><summary>test output</summary>
<p>

```
1) Scenario: Ability to specify multiple formatters # features/language.feature:17
   ✔ Before # features/support/hooks.js:21
   ✔ Given a file named "features/a.feature" with: # features/step_definitions/file_steps.js:11
       """
       Fonctionnalité: Bonjour
         Scénario: Monde
           Soit une étape
       """
   ✔ And a file named "features/step_definitions/cucumber_steps.js" with: # features/step_definitions/file_steps.js:11
       """
       import {Given} from 'cucumber'

       Given(/^une étape$/, function() {})
       """
   ✔ When I run cucumber-js with `--language fr` # features/step_definitions/cli_steps.js:9
   ✖ Then it outputs the text: # features/step_definitions/cli_steps.js:36
       """
       .

       1 scenario (1 passed)
       1 step (1 passed)
       <duration-stat>
       """
       AssertionError
           + expected - actual

           +.
           +
           +1 scenario (1 passed)
           +1 step (1 passed)
           +<duration-stat>
       
           at World.eql (/home/useloom/repos/cucumber-js/features/step_definitions/cli_steps.js:39:27)
   ✖ After # features/support/hooks.js:76
       Error: Last run errored unexpectedly. Output:



       Error Output:

       TypeError: Cannot read property 'line' of undefined
           at Object.generateEvents (/home/useloom/repos/cucumber-js/node_modules/gherkin/lib/gherkin/generate_events.js:61:38)
           at /home/useloom/repos/cucumber-js/lib/cli/helpers.js:111:37
           at Generator.next (<anonymous>)
           at asyncGeneratorStep (/home/useloom/repos/cucumber-js/lib/cli/helpers.js:29:103)
           at _next (/home/useloom/repos/cucumber-js/lib/cli/helpers.js:31:194)
           at /home/useloom/repos/cucumber-js/lib/cli/helpers.js:31:364
           at new Promise (<anonymous>)
           at /home/useloom/repos/cucumber-js/lib/cli/helpers.js:31:97
           at getTestCases (/home/useloom/repos/cucumber-js/lib/cli/helpers.js:97:24)
           at /home/useloom/repos/cucumber-js/lib/cli/helpers.js:77:39
           at Generator.next (<anonymous>)
           at asyncGeneratorStep (/home/useloom/repos/cucumber-js/lib/cli/helpers.js:29:103)
           at _next (/home/useloom/repos/cucumber-js/lib/cli/helpers.js:31:194)
           at <anonymous>
           at World.<anonymous> (/home/useloom/repos/cucumber-js/features/support/hooks.js:78:11)


216 scenarios (1 failed, 215 passed)
1067 steps (1 failed, 1066 passed)
1m08.437s
error Command failed with exit code 1.
```

</p>
</details>

* That was happening before I made any change to the code. I am unsure of the cause, (may be a problem with my local set up?)

### CODE REVIEW

File: features/skipped_steps.feature [hunk 1]
```diff
@@ -26,6 +26,14 @@ Feature: Skipped steps
...
     Then it passes
     And the step "a skipped step" has status "skipped"
+    And it outputs the text:
+      """
+      -
+
+      1 scenario (1 skipped)
+      1 step (1 skipped)
+      <duration-stat>
+      """
```
Comment (charlierudolph) [inline features/skipped_steps.feature]: Rather than testing the output, I'd prefer we create a step like `And the scenario "a scenario" has status "skipped"`. You should be able to model that step after similar steps in `features/step_definitions/json_output_steps.js`

File: src/runtime/test_case_runner.js [hunk 2]
```diff
@@ -127,17 +127,15 @@ export default class TestCaseRunner {
 
   shouldUpdateStatus(testStepResult) {
     switch (testStepResult.status) {
+      case Status.UNDEFINED:
```
Comment (charlierudolph) [inline src/runtime/test_case_runner.js]: Is this change needed to keep the tests passing after the change below?

If we are making this change, I'd prefer we also update the return for these three cases to be
```javascript
return (
  this.result.status !== Status.FAILED ||
  this.result.status !== Status.AMBIGUOUS || 
  this.result.status !== Status.UNDEFINED 
)
```
It would probably also be good to refactor this to use lodash's [_.some](https://lodash.com/docs/4.17.11#some)

File: features/skipped_steps.feature (hunk 1 above)
Comment (mgonnet) [reply features/skipped_steps.feature]: I agree that your suggestion would be more cleaner, and it was my first approach. However, when I checked `this.lastRun.jsonOuput` (or any json output for that matter) I found that the result of the Scenario itself is not stored, only the result of individual steps. An example below

<details><summary>output</summary>
<p>

```json
{
  "errorOutput": "",
  "jsonOutput": [
    {
      "keyword": "Feature",
      "name": "a feature",
      "line": 1,
      "id": "a-feature",
      "tags": [],
      "uri": "features/skipped.feature",
      "elements": [
        {
          "id": "a-feature;a-scenario",
          "keyword": "Scenario",
          "line": 2,
          "name": "a scenario",
          "tags": [],
          "type": "scenario",
          "steps": [
            {
              "keyword": "Before",
              "hidden": true,
              "match": {
                "location": "features/support/hooks.js:3"
              },
              "result": {
                "status": "skipped"
              }
            },
            {
              "arguments": [],
              "keyword": "Given ",
              "line": 3,
              "name": "a skipped step",
              "match": {
                "location": "features/step_definitions/skipped_steps.js:3"
              },
              "result": {
                "status": "skipped"
              }
            }
          ]
        }
      ]
    }
  ],
  "output": "--\n\n1 scenario (1 skipped)\n1 step (1 skipped)\n0m00.000s\n"
}
```

</p>
</details>

I am not sure of how to proceed or if I missing something. If needed, I can take a look and add the scenario result to the json output.

Comment (mgonnet) [reply features/skipped_steps.feature]: @charlierudolph

File: src/runtime/test_case_runner.js (hunk 2 above)
Comment (mgonnet) [reply src/runtime/test_case_runner.js]: @charlierudolph

File: features/skipped_steps.feature (hunk 1 above)
Comment (charlierudolph) [reply features/skipped_steps.feature]: Sorry, after seeing this I started work on making this possible by switching out parsing the json output for event protocol output. Haven't finish that yet though. I'm fine with this as is and will update the test once I complete my task.

**Comment by aslakhellesoy** [comment]
Hi @mgonnet,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by aslakhellesoy** [comment]
Hi @mgonnet,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber
